cdap-common/src/main/java/co/cask/cdap/common/guice/LocationRuntimeModule.java

```diff
@@ -85,40 +85,21 @@ private RootLocationFactory providesRootLocationFactory(CConfiguration cConf) {
     @Override
     protected void configure() {
       bind(NamespacedLocationFactory.class).to(DefaultNamespacedLocationFactory.class);
+      bind(RootLocationFactory.class).toProvider(RootLocationFactoryProvider.class).in(Scopes.SINGLETON);
 
       expose(LocationFactory.class);
       expose(RootLocationFactory.class);
       expose(NamespacedLocationFactory.class);
     }
 
-    @Provides
-    @Singleton
-    private RootLocationFactory provideRoot(Configuration hConf) {
-      return new RootLocationFactory(new FileContextLocationFactory(hConf));
-    }
 
     @Provides
     @Singleton
     private LocationFactory providesLocationFactory(final Configuration hConf, CConfiguration cConf) {
       final String namespace = cConf.get(Constants.CFG_HDFS_NAMESPACE);
       LOG.info("HDFS namespace is " + namespace);
 
-      String hdfsUser = cConf.get(Constants.CFG_HDFS_USER);
-      UserGroupInformation ugi;
-      try {
-        if (hdfsUser == null || UserGroupInformation.isSecurityEnabled()) {
-          if (hdfsUser != null) {
-            LOG.debug("Ignoring configuration {}={}, running on secure Hadoop", Constants.CFG_HDFS_USER, hdfsUser);
-          }
-          LOG.debug("Getting filesystem for current user");
-          ugi = UserGroupInformation.getCurrentUser();
-        } else {
-          LOG.debug("Getting filesystem for user {}", hdfsUser);
-          ugi = UserGroupInformation.createRemoteUser(hdfsUser);
-        }
-      } catch (Exception e) {
-        throw Throwables.propagate(e);
-      }
+      UserGroupInformation ugi = RootLocationFactoryProvider.getUserGroupInformation(cConf);
```

Comment:
After looking closer of this PR, you actually don't need to expose this method. You can have a private binding of `FileContext` so that it can be used by both `RootLocationFactory` and `LocationFactory`. E.g. something like this:

```java
class FileContextProvider implements Provider<FileContext> {
  private final CConfiguration cConf;
  private final Configuration hConf;
  
  @Inject
  FileContextProvider(CConfiguration cConf, Configuration hConf) {
    ...
  }

  @Override
  public FileContext get() {
    createUGI().doAs(new PrivilegedAction<FileContext>() {
      @Override
       public FileContext run() {
         return FileContext.getFileContext(hConf);
       }
    });
  }

  private UserGroupInformation createUGI() {
    String hdfsUser = cConf.get(Constants.CFG_HDFS_USER);
    // This is your getUserGroupInformation method implementation
    ...
  }
}
```

Then the RootLocationFactory provider can simply be
```java
  class RootLocationFactoryProvider implements Provider<RootLocationFactory> {
    private final FileContext fc;
    private final Configuration hConf;
    ....

    public RootLocationFactory get() {
      return new RootLocationFactory(new FileContextLocationFactory(hConf, fc, "/"));
    }
```
And the normal `LocationFactory` binding is
```java
    @Provides
    @Singleton
    private LocationFactory provideRoot(CConfiguration cConf, Configuration hConf, FileContext fc) {
      String namespace = cConf.get(Constants.CFG_HDFS_NAMESPACE);
      return new FileContextLocationFactory(new FileContextLocationFactory(hConf, fc, namespace));
    }
```